### DISCUSSION TITLE
How does Apicurio registry pick schema version if more than 1 version is uploaded

### DISCUSSION THREAD

**Discussion by junaidaj86** [Apicurio/apicurio-registry#2894] [Q&A] [answered]
Hello Team,

I am exploring Apicurio and had a query, I have a JSON schema which is uploaded in Apicurio version 2.x.
I have multiple version of the schema with little modification uploaded.
How does Apicurio pick which version of schema to validate with.
Also how do we set our application to use backward and forward compatibility feature.

Note: I am using spring boot, Kafka and apicurio registry 2.x with JSON validation

**Comment by EricWittmann** [comment, marked as answer]
Hi @junaidaj86 sorry for the late response.

A few things to answer here.

First, Apicurio Registry does not perform validation on the server.  Validation always occurs on the client side.  Typically that means the Serializer/Deserializer code in a typical Kafka application.

The way this usually works is that the **producer** Kafka application uses a particular schema version to produce a message on a Kafka topic.  The unique ID of that specific schema version is included in the Kafka message, either as a header or as the first 4 or 8 bytes of the message payload.  Next, when a **consumer** Kafka application reads the message, it determines the unique schema ID by reading it from the message.  It then uses that unique ID to lookup the schema version in the registry.  Then it uses that schema to either deserialize or validate the message body.

The question then is how does the **producer** application know what version of a schema to use.  That depends on how you want to do it.  

* You can bundle the schema itself with the application.  In this case the **producer** application is then responsible for ensuring that the schema it's using is registered in the registry.  Our Apicurio Registry Serializer class can be configured to do this automatically.
* You have some CI/CD or build process that registers the schema and then configures the application with the unique registry coordinates of the schema.  In this case the application would pull the schema from the registry at runtime and use it during serialization.

As far as the Apicurio Registry server is concerned, every version of a schema is unique and has unique coordinates.  It can be identified in one of these ways:

* Registry coordinates:  Group + Artifact Id + Version
* Global ID (8 byte unique number identifying a single version of an artifact)
* Content ID (8 byte number identifying a piece of content - this might be shared across multiple artifacts or versions if the exact same byte-identical content is uploaded multiple times as different artifacts or versions)
* Content Hash (SHA hash code identifying a piece of content)

Every registry artifact may consist of multiple versions.  The `latest` version of an artifact is simply the one that was most recently added.

**Comment by Surya-B-2243** [comment]
Hey @EricWittmann . Great explanation, what If I want to enforce a particular version on the client side (on the deserializer) . Instead of using the global id coming in the serialized message.
For example : I am receiving the serialized message in V1  schema. But I want to enforce the deserializer to use the latest version of the schema lets say V3 . How can I do this ?

**Reply by EricWittmann** [reply]
I don't think our Deserializer classes have any native configuration to support that.  However, you could use either the custom resolver strategy **or** the custom schema resolver interfaces that we expose.  We have examples for both:

## Resolver Strategy
The resolver strategy is responsible for figuring out the registry coordinates of the schema to use.  Typically the coordinates are extracted from the message (either the body or the headers).  But you could implement whatever you wanted.

* **Example**: https://github.com/Apicurio/apicurio-registry/tree/main/examples/custom-strategy/src/main/java/io/apicurio/registry/examples/custom/strategy

## Schema Resolver
The resolver is responsible for making a GET call to the Schema Registry to actually retrieve the content of the Schema to use.  Typically this uses the Apicurio Registry SDK to fetch the schema at the coordinates produced by the resolver strategy.  But again, you could provide your own implementation and do whatever you wanted.

*  **Example**: https://github.com/Apicurio/apicurio-registry/tree/main/examples/custom-resolver/src/main/java/io/apicurio/registry/examples/custom/resolver
